File: integration/ts/covenant-vault.test.ts
/**
 * CovenantVault integration test — stateless contract with checkSig + checkPreimage.
 *
 * ## How It Works
 *
 * CovenantVault demonstrates a covenant pattern: it constrains HOW funds can be spent,
 * not just WHO can spend them. The contract checks:
 *   1. The owner's ECDSA signature (authentication via checkSig)
 *   2. The transaction preimage (via checkPreimage, which enables script-level
 *      inspection of the spending transaction)
 *   3. That the spending amount >= minAmount (covenant rule)
 *
 * ### What is checkPreimage / OP_PUSH_TX?
 *   checkPreimage verifies a BIP-143 sighash preimage against the spending transaction.
 *   This is implemented via the OP_PUSH_TX technique: the unlocking script pushes
 *   both a preimage (the raw BIP-143 serialization) and an ECDSA signature computed
 *   with private key k=1 (whose public key is the generator point G). The locking
 *   script verifies this signature against the preimage, which proves the preimage
 *   is genuine. Once verified, the script can inspect transaction fields (outputs,
 *   amounts, etc.) by parsing the preimage — enabling covenant rules.
 *
 * ### Constructor
 *   - owner: PubKey — the ECDSA public key that must sign to spend
 *   - recipient: Addr — the hash160 of the authorized recipient's public key
 *   - minAmount: bigint — minimum satoshis that must be sent to the recipient
 *
 * ### Method: spend(sig: Sig, amount: bigint, txPreimage: SigHashPreimage)
 *   The compiler inserts an implicit _opPushTxSig parameter before the declared params.
 *   The full unlocking script order is: <opPushTxSig> <sig> <amount> <txPreimage>
 *
 *   - sig: owner's ECDSA signature (auto-computed by SDK when null)
 *   - amount: satoshis to send to recipient (must be >= minAmount)
 *   - txPreimage: BIP-143 sighash preimage (auto-computed by SDK when null)
 *
 * ### SDK Auto-Compute
 *   The SDK's call() method detects SigHashPreimage params set to null and
 *   automatically computes the OP_PUSH_TX signature and BIP-143 preimage.
 *   It also auto-computes Sig params set to null using the signer's private key.
 *   This means the caller only needs to provide the amount — both cryptographic
 *   values are computed from the spending transaction.
 */

import { describe, it, expect } from 'vitest';
import { compileContract } from './helpers/compile.js';
import { RunarContract, RPCProvider } from 'runar-sdk';
import { createFundedWallet, createWallet } from './helpers/wallet.js';

function createProvider() {
  return new RPCProvider('http://localhost:18332', 'regtest', 'regtest', {
    autoMine: true,
    network: 'testnet',
  });
}

describe('CovenantVault', () => {
  it('should compile the CovenantVault contract', () => {
    const artifact = compileContract('examples/ts/covenant-vault/CovenantVault.runar.ts');
    expect(artifact).toBeTruthy();
    expect(artifact.contractName).toBe('CovenantVault');
  });

  it('should deploy with owner, recipient, and minAmount', async () => {
    const artifact = compileContract('examples/ts/covenant-vault/CovenantVault.runar.ts');

    const provider = createProvider();
    const owner = createWallet();
    const recipient = createWallet();
    const { signer } = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [
      owner.pubKeyHex,
      recipient.pubKeyHash,
      1000n,
    ]);

    const { txid: deployTxid } = await contract.deploy(provider, signer, { satoshis: 5000 });
    expect(deployTxid).toBeTruthy();
    expect(typeof deployTxid).toBe('string');
    expect(deployTxid.length).toBe(64);
  });

  it('should deploy with zero minAmount', async () => {
    const artifact = compileContract('examples/ts/covenant-vault/CovenantVault.runar.ts');

    const provider = createProvider();
    const owner = createWallet();
    const recipient = createWallet();
    const { signer } = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [
      owner.pubKeyHex,
      recipient.pubKeyHash,
      0n,
    ]);

    const { txid: deployTxid } = await contract.deploy(provider, signer, { satoshis: 5000 });
    expect(deployTxid).toBeTruthy();
  });

  it('should deploy with large minAmount', async () => {
    const artifact = compileContract('examples/ts/covenant-vault/CovenantVault.runar.ts');

    const provider = createProvider();
    const owner = createWallet();
    const recipient = createWallet();
    const { signer } = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [
      owner.pubKeyHex,
      recipient.pubKeyHash,
      100_000_000n, // 1 BTC in satoshis
    ]);

    const { txid: deployTxid } = await contract.deploy(provider, signer, { satoshis: 5000 });
    expect(deployTxid).toBeTruthy();
  });

  it('should deploy with same key as owner and recipient', async () => {
    const artifact = compileContract('examples/ts/covenant-vault/CovenantVault.runar.ts');

    const provider = createProvider();
    const ownerAndRecipient = createWallet();
    const { signer } = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [
      ownerAndRecipient.pubKeyHex,
      ownerAndRecipient.pubKeyHash,
      500n,
    ]);

    const { txid: deployTxid } = await contract.deploy(provider, signer, { satoshis: 5000 });
    expect(deployTxid).toBeTruthy();
  });

  /**
   * Deploy and spend with valid owner signature, preimage, and amount >= minAmount.
   *
   * Steps:
   *   1. Create owner wallet (will be the signer — their ECDSA key must match the constructor)
   *   2. Deploy with (ownerPubKey, recipientPubKeyHash, minAmount=1000)
   *   3. Call spend(null, 2000n, null):
   *      - null Sig → SDK auto-computes ECDSA signature from signer's private key
   *      - 2000n   → amount (>= minAmount of 1000)
   *      - null SigHashPreimage → SDK auto-computes BIP-143 preimage and _opPushTxSig
   *   4. The SDK builds the unlocking script: <opPushTxSig> <sig> <amount> <txPreimage>
   *   5. On-chain, the script verifies checkSig(sig, owner), checkPreimage(txPreimage),
   *      and asserts amount >= minAmount
   */
  it('should spend with valid signature and amount >= minAmount', async () => {
    const artifact = compileContract('examples/ts/covenant-vault/CovenantVault.runar.ts');

    const provider = createProvider();
    const recipient = createWallet();

    // Owner must be the signer — their ECDSA key must match constructor's owner param
    const { signer: ownerSigner, pubKeyHex: ownerPubKeyHex, pubKeyHash: recipientPKH } = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [
      ownerPubKeyHex,
      recipient.pubKeyHash,
      1000n, // minAmount
    ]);

    await contract.deploy(provider, ownerSigner, { satoshis: 5000 });

    // spend(sig=null, amount=2000, txPreimage=null)
    // SDK auto-computes both Sig and SigHashPreimage from the spending transaction
    const { txid: spendTxid } = await contract.call(
      'spend',
      [null, 2000n, null],
      provider,
      ownerSigner,
    );
    expect(spendTxid).toBeTruthy();
    expect(typeof spendTxid).toBe('string');
    expect(spendTxid.length).toBe(64);
  });

  it('should reject spend with amount below minAmount', async () => {
    const artifact = compileContract('examples/ts/covenant-vault/CovenantVault.runar.ts');

    const provider = createProvider();
    const recipient = createWallet();

    // Owner must be the signer
    const { signer: ownerSigner, pubKeyHex: ownerPubKeyHex } = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [
      ownerPubKeyHex,
      recipient.pubKeyHash,
      1000n, // minAmount=1000
    ]);

    await contract.deploy(provider, ownerSigner, { satoshis: 5000 });

    // spend(sig=null, amount=500, txPreimage=null) — amount < minAmount should fail
    await expect(
      contract.call('spend', [null, 500n, null], provider, ownerSigner),
    ).rejects.toThrow();
  });
});

File: integration/go/covenant_vault_test.go
//go:build integration

package integration

import (
	"encoding/hex"
	"testing"

	"runar-integration/helpers"

	"github.com/bsv-blockchain/go-sdk/script"

	runar "github.com/icellan/runar/packages/runar-go"
)

func deployCovenantVault(t *testing.T, owner, recipient *helpers.Wallet, minAmount int64) *runar.RunarContract {
	t.Helper()

	artifact, err := helpers.CompileToSDKArtifact(
		"examples/ts/covenant-vault/CovenantVault.runar.ts",
		map[string]interface{}{},
	)
	if err != nil {
		t.Fatalf("compile: %v", err)
	}
	t.Logf("CovenantVault script: %d bytes", len(artifact.Script)/2)

	contract := runar.NewRunarContract(artifact, []interface{}{
		owner.PubKeyHex(),
		recipient.PubKeyHashHex(),
		int64(minAmount),
	})

	funder := helpers.NewWallet()
	helpers.RPCCall("importaddress", funder.Address, "", false)
	_, err = helpers.FundWallet(funder, 1.0)
	if err != nil {
		t.Fatalf("fund: %v", err)
	}

	provider := helpers.NewRPCProvider()
	signer, err := helpers.SDKSignerFromWallet(funder)
	if err != nil {
		t.Fatalf("signer: %v", err)
	}

	_, _, err = contract.Deploy(provider, signer, runar.DeployOptions{Satoshis: 5000})
	if err != nil {
		t.Fatalf("deploy: %v", err)
	}

	return contract
}

func TestCovenantVault_Compile(t *testing.T) {
	artifact, err := helpers.CompileToSDKArtifact(
		"examples/ts/covenant-vault/CovenantVault.runar.ts",
		map[string]interface{}{},
	)
	if err != nil {
		t.Fatalf("compile: %v", err)
	}
	if artifact.ContractName != "CovenantVault" {
		t.Fatalf("expected contract name CovenantVault, got %s", artifact.ContractName)
	}
	t.Logf("CovenantVault compiled: %d bytes", len(artifact.Script)/2)
}

func TestCovenantVault_Deploy(t *testing.T) {
	owner := helpers.NewWallet()
	recipient := helpers.NewWallet()
	contract := deployCovenantVault(t, owner, recipient, 1000)
	utxo := contract.GetCurrentUtxo()
	if utxo == nil {
		t.Fatalf("no UTXO after deploy")
	}
	t.Logf("deployed with minAmount=1000")
}

func TestCovenantVault_DeployZeroMinAmount(t *testing.T) {
	owner := helpers.NewWallet()
	recipient := helpers.NewWallet()
	contract := deployCovenantVault(t, owner, recipient, 0)
	utxo := contract.GetCurrentUtxo()
	if utxo == nil {
		t.Fatalf("no UTXO after deploy")
	}
	t.Logf("deployed with minAmount=0")
}

func TestCovenantVault_DeployLargeMinAmount(t *testing.T) {
	owner := helpers.NewWallet()
	recipient := helpers.NewWallet()
	contract := deployCovenantVault(t, owner, recipient, 100000000)
	utxo := contract.GetCurrentUtxo()
	if utxo == nil {
		t.Fatalf("no UTXO after deploy")
	}
	t.Logf("deployed with minAmount=100000000 (1 BTC)")
}

func TestCovenantVault_DeploySameKey(t *testing.T) {
	wallet := helpers.NewWallet()
	contract := deployCovenantVault(t, wallet, wallet, 1000)
	utxo := contract.GetCurrentUtxo()
	if utxo == nil {
		t.Fatalf("no UTXO after deploy")
	}
	t.Logf("deployed with same key as owner and recipient")
}

func TestCovenantVault_ValidSpend(t *testing.T) {
	owner := helpers.NewWallet()
	recipient := helpers.NewWallet()
	minAmount := int64(1000)

	contract := deployCovenantVault(t, owner, recipient, minAmount)

	// Get UTXO from SDK contract, convert for raw spending
	utxo := helpers.SDKUtxoToHelper(contract.GetCurrentUtxo())

	// Build spend TX
	receiverScript := owner.P2PKHScript()
	spendTx, err := helpers.BuildSpendTx(utxo, receiverScript, 4500)
	if err != nil {
		t.Fatalf("build spend: %v", err)
	}

	// Sign with owner's key
	sigHex, err := helpers.SignInput(spendTx, 0, owner.PrivKey)
	if err != nil {
		t.Fatalf("sign: %v", err)
	}
	sigBytes, _ := hex.DecodeString(sigHex)

	// OP_PUSH_TX for checkPreimage
	opPushTxSigHex, preimageHex, err := helpers.SignOpPushTx(spendTx, 0)
	if err != nil {
		t.Fatalf("op_push_tx: %v", err)
	}
	opPushTxSigBytes, _ := hex.DecodeString(opPushTxSigHex)
	preimageBytes, _ := hex.DecodeString(preimageHex)

	// spend(sig: Sig, amount: bigint, txPreimage: SigHashPreimage)
	// Compiler inserts implicit _opPushTxSig before declared params.
	// Unlocking script order: <opPushTxSig> <sig> <amount> <txPreimage>
	amount := int64(2000) // >= minAmount (1000)
	unlockHex := helpers.EncodePushBytes(opPushTxSigBytes) +
		helpers.EncodePushBytes(sigBytes) +
		helpers.EncodePushInt(amount) +
		helpers.EncodePushBytes(preimageBytes)

	unlockScript, _ := script.NewFromHex(unlockHex)
	spendTx.Inputs[0].UnlockingScript = unlockScript

	txid := helpers.AssertTxAccepted(t, spendTx.Hex())
	helpers.AssertTxInBlock(t, txid)
}

func TestCovenantVault_BelowMinAmount_Rejected(t *testing.T) {
	owner := helpers.NewWallet()
	recipient := helpers.NewWallet()
	minAmount := int64(1000)

	contract := deployCovenantVault(t, owner, recipient, minAmount)

	utxo := helpers.SDKUtxoToHelper(contract.GetCurrentUtxo())

	spendTx, err := helpers.BuildSpendTx(utxo, owner.P2PKHScript(), 4500)
	if err != nil {
		t.Fatalf("build spend: %v", err)
	}

	sigHex, err := helpers.SignInput(spendTx, 0, owner.PrivKey)
	if err != nil {
		t.Fatalf("sign: %v", err)
	}
	sigBytes, _ := hex.DecodeString(sigHex)

	opPushTxSigHex, preimageHex, err := helpers.SignOpPushTx(spendTx, 0)
	if err != nil {
		t.Fatalf("op_push_tx: %v", err)
	}
	opPushTxSigBytes, _ := hex.DecodeString(opPushTxSigHex)
	preimageBytes, _ := hex.DecodeString(preimageHex)

	// Amount 500 < minAmount 1000 — should fail assert
	// Unlocking script order: <opPushTxSig> <sig> <amount> <txPreimage>
	amount := int64(500)
	unlockHex := helpers.EncodePushBytes(opPushTxSigBytes) +
		helpers.EncodePushBytes(sigBytes) +
		helpers.EncodePushInt(amount) +
		helpers.EncodePushBytes(preimageBytes)

	unlockScript, _ := script.NewFromHex(unlockHex)
	spendTx.Inputs[0].UnlockingScript = unlockScript

	helpers.AssertTxRejected(t, spendTx.Hex())
}
